asmcomp/CSEgen.ml

```diff
@@ -310,59 +310,76 @@ method private cse n i =
+         self#cse n2 i.next (fun next -> k { i with next; })
       | Op_store true ->
           (* A non-initializing store can invalidate
              anything we know about prior loads. *)
          let n1 = set_unknown_regs n (Proc.destroyed_at_oper i.desc) in
          let n2 = set_unknown_regs n1 i.res in
          let n3 = self#kill_loads n2 in
-         {i with next = self#cse n3 i.next}
+         self#cse n3 i.next (fun next -> k { i with next; })
       end
   (* For control structures, we set the numbering to empty at every
      join point, but propagate the current numbering across fork points. *)
   | Iifthenelse(test, ifso, ifnot) ->
-     let n1 = set_unknown_regs n (Proc.destroyed_at_oper i.desc) in
-      {i with desc = Iifthenelse(test, self#cse n1 ifso, self#cse n1 ifnot);
-              next = self#cse empty_numbering i.next}
+      let n1 = set_unknown_regs n (Proc.destroyed_at_oper i.desc) in
+      self#cse n1 ifso (fun ifso ->
+        self#cse n1 ifnot (fun ifnot ->
+          self#cse empty_numbering i.next (fun next ->
+            k { i with desc = Iifthenelse(test, ifso, ifnot); next; })))
   | Iswitch(index, cases) ->
-     let n1 = set_unknown_regs n (Proc.destroyed_at_oper i.desc) in
-      {i with desc = Iswitch(index, Array.map (self#cse n1) cases);
-              next = self#cse empty_numbering i.next}
+      let n1 = set_unknown_regs n (Proc.destroyed_at_oper i.desc) in
+      self#cse_array n1 cases (fun cases ->
+        self#cse empty_numbering i.next (fun next ->
+          k { i with desc = Iswitch(index, cases); next; }))
   | Icatch(rec_flag, handlers, body) ->
-      let aux (nfail, handler) =
-        nfail, self#cse empty_numbering handler
-      in
-      {i with desc = Icatch(rec_flag, List.map aux handlers, self#cse n body);
-              next = self#cse empty_numbering i.next}
+      let nfail, handler_code = List.split handlers in
+      self#cse_list empty_numbering handler_code (fun handler_code ->
+        let handlers = List.combine nfail hander_code in
+        self#cse n body (fun body ->
+          self#cse empty_numbering i.next (fun next ->
+            k { i with desc = Icatch(rec_flag, handlers, body); next; })))
   | Itrywith(body, handler) ->
-      {i with desc = Itrywith(self#cse n body,
-                              self#cse empty_numbering handler);
-              next = self#cse empty_numbering i.next}
+      self#cse n body (fun body ->
+        self#cse empty_numbering handler (fun handler ->
+          self#cse empty_numbering i.next (fun next ->
+            k { i with desc = Itrywith(body, handler); next; })))
+
+method private cse_array n is k =
+  self#cse_list n (Array.to_list is) (fun is -> k (Array.of_list is))
+
+method private cse_list0 n is acc k =
+  match is with
+  | [] -> k acc
+  | i::is -> self#cse n i (fun i -> self#cse_list0 n is (i :: acc) k)
+
+method private cse_list n is k =
+  self#cse_list0 n is [] (fun is_rev -> k (List.rev is_rev))
```

Comment:
```suggestion
method private cse_list0 n is acc k =
  match is with
  | [] -> k (List.rev acc)
  | i::is -> self#cse n i (fun i -> self#cse_list0 n is (i :: acc) k)

method private cse_list n is k =
  self#cse_list0 n is [] k
```
Make it obvious that `acc` is in reverse order, and at the same time make that fact private to `cse_list0`.